File: src/main/java/com/wesker926/leetcode/algorithms/p1005/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/12/3
// @description: 贪心 + 桶排序
package s1

func largestSumAfterKNegations(nums []int, k int) int {
	bucket := make([]int, 201)
	for _, num := range nums {
		bucket[num+100]++
	}

	sum, min := 0, 100
	for i, c := range bucket {
		if c == 0 {
			continue
		}

		num := i - 100
		if num >= 0 {
			sum += num * c
		} else {
			if k >= c {
				sum += (-num) * c
				k -= c
			} else {
				sum += (-num) * k
				sum += num * (c - k)
				k = 0
			}
			num = -num
		}

		if num < min {
			min = num
		}
	}

	if k&1 == 0 {
		return sum
	}
	return sum - min*2
}

File: src/main/java/com/wesker926/leetcode/algorithms/p1005/s1/Solution.java
package com.wesker926.leetcode.algorithms.p1005.s1;

/**
 * @author wesker.gh
 * @date 2021/12/3
 * @description 贪心 + 桶排序
 */
public class Solution {
    public int largestSumAfterKNegations(int[] nums, int k) {
        int[] bucket = new int[201];
        for (int num : nums) {
            bucket[num + 100]++;
        }

        int sum = 0, min = 100, num, count;
        for (int i = 0; i < bucket.length; i++) {
            if (bucket[i] == 0) {
                continue;
            }

            count = bucket[i];
            num = i - 100;
            if (num >= 0) {
                sum += num * count;
            } else {
                if (k >= count) {
                    sum += (-num) * count;
                    k -= count;
                } else {
                    sum += (-num) * k;
                    sum += num * (count - k);
                    k = 0;
                }
                num = -num;
            }
            min = Math.min(min, num);
        }

        return (k & 1) == 0 ? sum : sum - min * 2;
    }
}
